heroic-core/src/main/java/com/spotify/heroic/cluster/CoreClusterManager.java

```diff
@@ -276,30 +275,32 @@ public void register(LifeCycleRegistry registry) {
             throw new IllegalStateException("Registry not ready");
         }
 
-        final List<Pair<Map<String, String>, List<ClusterNode>>> shards =
+        List<Pair<Map<String, String>, List<ClusterNode>>> shards =
             registry.findFromAllShards(OptionalLimit.empty());
 
-        if (expectedShards.isPresent()) {
-            final Set<Map<String, String>> foundShards = new HashSet<>(shards
-                .stream()
-                .map(mapListPair -> mapListPair.getKey())
-                .collect(Collectors.toList()));
-
-            final List<Map<String, String>> shardsWithNoNodes = new ArrayList<>();
-            for (final Map<String, String> expected : expectedShards.get()) {
-                if (!foundShards.contains(expected)) {
-                    shardsWithNoNodes.add(expected);
-                }
-            }
-
-            if (shardsWithNoNodes.size() > 0) {
-                for (final Map<String, String> shard : shardsWithNoNodes) {
-                    /* For every shard that didn't have a discovered node, add an empty entry in the
-                     * shard list. This ensures that suitable code later on will complain that there
-                     * were shards with no available nodes/groups. */
-                    shards.add(Pair.of(shard, ImmutableList.of()));
-                }
+        /* Topology (shards) is detected based on the metadata coming from the nodes. Any shards
+         * mentioned in 'topology' are required (shard error is raised if it isn't), additional
+         * shards may also exist tho. */
+        final Set<Map<String, String>> foundShards =
+            new HashSet<>(shards.stream().map(Pair::getKey).collect(Collectors.toList()));
+
+        log.info("expectedTopology: " + expectedTopology);
+
+        final List<Map<String, String>> shardsWithNoNodes = expectedTopology
+            .stream()
+            .filter(e -> !foundShards.contains(e))
+            .collect(Collectors.toList());
+
+        if (shardsWithNoNodes.size() > 0) {
+            List<Pair<Map<String, String>, List<ClusterNode>>> updated = new ArrayList<>();
+            updated.addAll(shards);
+            for (final Map<String, String> shard : shardsWithNoNodes) {
+                /* For every shard that didn't have a discovered node, add an empty entry in the
+                 * shard list. This ensures that suitable code later on will complain that there
+                 * were shards with no available nodes/groups. */
+                updated.add(Pair.of(shard, ImmutableList.of()));
             }
+            shards = updated;
```

Comment:
branch can be inverted, and immediately return `shards` (which can then be final). In the other case, just return updated.

This also has the benefit of the branch being smaller.

```
if (shardsWithNoNodes.isEmpty()) {
    return shards;
}

/* build and return updated */
```